### PR TITLE
Health check for stores

### PR DESCRIPTION
**Pull request by huachaohuang**
Filter out unknown and down stores from balance.

### CODE REVIEW

File: server/balancer.go [hunk 1]
```diff
@@ -127,7 +128,7 @@ func (cb *capacityBalancer) ScoreType() scoreType {
 // selectBalanceRegion tries to select a store leader region to do balance and returns true, but if we cannot find any,
 // we try to find a store follower region and returns false.
 func (cb *capacityBalancer) selectBalanceRegion(cluster *clusterInfo, stores []*storeInfo) (*metapb.Region, *metapb.Peer, *metapb.Peer, bool) {
-	store := selectFromStore(stores, cluster.getUnknownStores(), cb.filters, cb.st)
+	store := selectFromStore(stores, nil, cb.filters, cb.st)
```
Comment (siddontang) [inline server/balancer.go]: why nil now?

Comment (huachaohuang) [inline server/balancer.go]: Unknown stores will be filtered out in `stateFilter`, no need to maintain another unknown stores map now.

Comment (huachaohuang) [inline server/balancer.go]: Some places still need this arg, we may remove it after we do some refactor, but not in this PR.

File: server/config.go [hunk 2]
```diff
@@ -187,6 +187,10 @@ type BalanceConfig struct {
 
 	// MaxTransferWaitCount is the max heartbeat count to wait leader transfer to finish.
 	MaxTransferWaitCount uint64 `toml:"max-transfer-wait-count" json:"max-transfer-wait-count"`
+
+	// MaxStoreDownInterval is the max interval (in seconds)
```
Comment (qiuyesuifeng) [inline server/config.go]: @queenypingcap  please help to check comments.

File: server/filter.go [hunk 3]
```diff
@@ -44,6 +46,35 @@ func filterToStore(store *storeInfo, filters []Filter, args ...interface{}) bool
...
+}
+
+func (sf *stateFilter) filterBadStore(store *storeInfo) bool {
+	if store.stats.Stats == nil {
+		// Unknown
+		return true
+	}
+	interval := time.Since(store.stats.LastHeartbeatTS).Seconds()
+	if uint64(interval) > sf.cfg.MaxStoreDownInterval {
+		// Down
```
Comment (qiuyesuifeng) [inline server/filter.go]: This comment is too short.

File: conf/config.toml [hunk 4]
```diff
@@ -25,4 +25,5 @@ balance-interval = 30
 max-balance-count = 16
 max-balance-retry-per-loop = 10
 max-balance-count-per-loop = 3
-max-transfer-wait-count = 3
\ No newline at end of file
+max-transfer-wait-count = 3
+max-store-down-interval = 60
```
Comment (qiuyesuifeng) [inline conf/config.toml]: Should we also put this config flag to main.go?

File: server/balancer_test.go [hunk 5]
```diff
@@ -340,3 +341,68 @@ func (s *testBalancerSuite) TestCapacityBalancer(c *C) {
...
+	leader := region.GetPeers()[0]
+
+	// Test add peer.
+	s.addRegionPeer(c, clusterInfo, 4, region, leader)
+
+	// Test add another peer.
+	s.addRegionPeer(c, clusterInfo, 3, region, leader)
+
+	// Make store 2 up and down.
+	for i := 0; i < 3; i++ {
```
Comment (qiuyesuifeng) [inline server/balancer_test.go]: Why using this loop for test? Seems one loop is enough.

Comment (huachaohuang) [inline server/balancer_test.go]: Make it up and down and up again for a few times.

File: server/config.go [hunk 6]
```diff
@@ -251,6 +256,10 @@ func (c *BalanceConfig) adjust() {
 	if c.MaxTransferWaitCount == 0 {
 		c.MaxTransferWaitCount = defaultMaxTransferWaitCount
 	}
+
+	if c.MaxStoreDownInterval == 0 {
+		c.MaxStoreDownInterval = defaultMaxStoreDownInterval
```
Comment (qiuyesuifeng) [inline server/config.go]: How about MaxStoreDownTime?

Comment (huachaohuang) [inline server/config.go]: I think `MaxStoreDownDuration` is better.

File: conf/config.toml (hunk 4 above)
Comment (huachaohuang) [inline conf/config.toml]: Yep, I miss it.

File: cmd/pd-server/main.go [hunk 7]
```diff
@@ -49,6 +49,7 @@ var (
 	maxBalanceRetryPerLoop = flag.Uint64("max-balance-retry-per-loop", 10, "the max retry count to balance in a balance schedule")
 	maxBalanceCountPerLoop = flag.Uint64("max-balance-count-per-loop", 3, "the max region count to balance in a balance schedule")
 	maxTransferWaitCount   = flag.Uint64("max-transfer-wait-count", 3, "the max heartbeat count to wait leader transfer to finish")
+	maxStoreDownDuration   = flag.Uint64("max-store-down-duration", 60, "the max duration a store without heartbeat will be considered to be down")
```
Comment (siddontang) [inline cmd/pd-server/main.go]: @queenypingcap  please review the explanation.

File: cmd/pd-server/main.go [hunk 8]
```diff
@@ -49,6 +49,7 @@ var (
 	maxBalanceRetryPerLoop = flag.Uint64("max-balance-retry-per-loop", 10, "the max retry count to balance in a balance schedule")
 	maxBalanceCountPerLoop = flag.Uint64("max-balance-count-per-loop", 3, "the max region count to balance in a balance schedule")
 	maxTransferWaitCount   = flag.Uint64("max-transfer-wait-count", 3, "the max heartbeat count to wait leader transfer to finish")
```
Comment (queenypingcap) [inline cmd/pd-server/main.go]: the max heartbeat count to wait for the leader transfer to finish

Comment (siddontang) [inline cmd/pd-server/main.go]: em, I want you see maxStoreDownDuration...

Comment (queenypingcap) [inline cmd/pd-server/main.go]: without heartbeat ->without heartbeats

**Comment by siddontang** [comment]
LGTM

**Comment by queenypingcap** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**